stonesoup/sensormanager/reinforcement.py

```diff
@@ -0,0 +1,294 @@
+        """Return action_spec."""
+        return self._action_spec
+
+    def observation_spec(self):
+        """Return observation_spec."""
+        return self._observation_spec
+
+
+class ReinforcementSensorManager(SensorManager):
+    """A sensor manager that employs reinforcement learning algorithms from tensorflow-agents.
+    The sensor manager trains on an environment to find an optimal policy, which is then exploited
+    to choose actions.
+    """
+    env: BaseEnvironment = Property(doc="The environment which the agent learns the policy with.")
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.tf_env = tf_py_environment.TFPyEnvironment(self.env)
+        self.test_env = tf_py_environment.TFPyEnvironment(self.env)
+        self.agent = None
+
+    @staticmethod
+    def compute_avg_return(environment, policy, num_episodes=10):
+        """Used to calculate the average reward over a set of episodes.
+
+        Parameters
+        ----------
+        environment:
+            tf-agents environment for evaluating policy on
+
+        policy:
+            tf-agents policy for choosing actions in environment
+
+        num_episodes: int
+            Number of episodes to sample over
+
+        Returns
+        -------
+        : int
+            average reward calculated over num_episodes
+
+        """
+        time_step = None
+        episode_return = None
+        total_return = 0.0
+        for _ in range(num_episodes):
+            time_step = environment.reset()
+            episode_return = 0.0
+
+            while not time_step.is_last():
+                action_step = policy.action(time_step)
+                time_step = environment.step(action_step.action)
+                episode_return += time_step.reward
+            total_return += episode_return
+
+        avg_return = total_return / num_episodes
+        return avg_return.numpy()[0]
+
+    @staticmethod
+    def dense_layer(num_units):
+        """Method for generating fully connected layers for use in the neural network.
+
+        Parameters
+        ----------
+        num_units: int
+            Number of nodes in dense layer
+
+        Returns
+        -------
+        : tensorflow dense layer
+
+        """
+        # Define a helper function to create Dense layers configured with the right
+        # activation and kernel initializer.
+        return tf.keras.layers.Dense(
+            num_units,
+            activation=tf.keras.activations.relu,
+            kernel_initializer=tf.keras.initializers.VarianceScaling(
+                scale=2.0, mode='fan_in', distribution='truncated_normal'))
+
+    def train(self, hyper_parameters):
+        """Trains a DQN agent on the specified environment to learn a policy that is later
+        used to select actions.
+
+        Parameters
+        ----------
+        hyper_parameters: dict
+            Dictionary containing hyperparameters used in training.
```

Comment:
What hyperparameters can be used/are needed? - Could just say see example/tutorial

Reply:
Added comment.